### PR TITLE
Add settings for write and erase pen radius.

### PR DESCRIPTION
**Pull request by bbirchnz**
Passed KneeboardState object to Tab so changes there will reflect across the board.
No UI yet.

Will need close review for code quality/conformance!

### CODE REVIEW

File: src/app/app-winui3/InputSettingsPage.xaml.cpp [hunk 1]
```diff
@@ -37,6 +37,17 @@ InputSettingsPage::InputSettingsPage() {
...
+winrt::event_token InputSettingsPage::PropertyChanged(
+  winrt::Microsoft::UI::Xaml::Data::PropertyChangedEventHandler const&
+    handler) {
+  return mPropertyChangedEvent.add(handler);
+}
+
+void InputSettingsPage::PropertyChanged(
+  winrt::event_token const& token) noexcept {
+  mPropertyChangedEvent.remove(token);
+}
```
Comment (fredemmott) [inline src/app/app-winui3/InputSettingsPage.xaml.cpp:49]: This is needed for two-way bindings, but I think you only need one-way bindings:
- AIUI the new UI you're adding is the only way to change the new events
- if there were another way, it doesn't appear that the PropertyChanged event is emitted by anything?

File: src/app/app-winui3/InputSettingsPage.xaml.cpp [hunk 2]
```diff
@@ -74,4 +85,37 @@ void InputSettingsPage::OnOrientationChanged(
...
+bool InputSettingsPage::FixedEraseRadius() {
+  return gKneeboard->GetDoodleSettings().FixedErase;
+}
+void InputSettingsPage::FixedEraseRadius(bool value) {
+  auto ds = gKneeboard->GetDoodleSettings();
+  ds.FixedErase = value;
+  gKneeboard->SetDoodleSettings(ds);
+}
+unsigned int InputSettingsPage::EraseRadius() {
```
Comment (fredemmott) [inline src/app/app-winui3/InputSettingsPage.xaml.cpp:104]: Use `uint32_t` (both here and in the header) to unambiguously match the IDL - though they're currently equivalent.

File: src/app/app-common/include/OpenKneeboard/DoodleSettings.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <nlohmann/json.hpp>
+
+namespace OpenKneeboard {
+struct DoodleSettings final {
+  uint32_t DrawRadius = 50;
+  uint32_t EraseRadius = 150;
+  bool FixedDraw = false;
+  bool FixedErase = true;
+  // TODO: stroke colour?
```
Comment (fredemmott) [inline src/app/app-common/include/OpenKneeboard/DoodleSettings.h:11]: Doesn't matter for a comment though FYI: despite growing up in England, I'm generally using US English here. DirectX and the windows API consistently use 'color', so using 'color' means we need to mix different spellings in the same code

**Review by fredemmott — CHANGES_REQUESTED** [review]
Thanks, this looks great!

Almost everything here is me being very picky about unimportant things. A few higher-level things:

* fixed width: is this better for you than drastically reduced sensitivity? I've found drastically reduced sensitivity has nearly the same effect, but having some pressure-based variance makes handwriting more natural and more readable
* having small unrelated formatting changes is fine, but it's best not to start autoformatting files that weren't previously formatted in the same PR/commit as non-formatting changes; a few reasons for this:
  * it reduces the changes of merge/rebase conflicts
  * if needed, it increases the chance of a revert working
  * it makes it easier to review the meaningful changes
  * if they are in separate commits, bulk autoformatting changes can be ignored by `git blame` with `--ignore-revs-file`, and [GitHub support for this is in beta](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)
  
Thanks again, this looks very very close to doable and is an amazing first PR

File: src/app/app-common/include/OpenKneeboard/DoodleSettings.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+};
+
+void from_json(const nlohmann::json&, DoodleSettings&);
+void to_json(nlohmann::json&, const DoodleSettings&);
```
[the remark this answers is not in the document]
Comment (bbirchnz) [reply src/app/app-common/include/OpenKneeboard/DoodleSettings.h:15]: Thanks, great to know!

File: src/app/app-winui3/InputDeviceUIData.cpp [hunk 5]
```diff
@@ -61,7 +61,8 @@ DataTemplate InputDeviceUIDataTemplateSelector::TabletDevice() {
   return mTabletDevice;
 }
 
-void InputDeviceUIDataTemplateSelector::TabletDevice(const DataTemplate& value) {
+void InputDeviceUIDataTemplateSelector::TabletDevice(
+  const DataTemplate& value) {
```
[the remark this answers is not in the document]
Comment (bbirchnz) [reply src/app/app-winui3/InputDeviceUIData.cpp:65]: Noted for future, this file had lots of things added and removed and I checked it in by mistake. I can't get clang-format to un-do this now though

File: src/app/app-winui3/InputSettingsPage.idl [hunk 6]
```diff
@@ -1,30 +1,33 @@
...
+  Windows.Foundation.Collections.IVector<IInspectable> Devices {
+    get;
+  };
+  Boolean FixedWriteRadius;
+  Boolean FixedEraseRadius;
+  UInt32 WriteRadius;
+  UInt32 EraseRadius;
 }
+
+}// namespace OpenKneeboardApp
```
[the remark this answers is not in the document]
Comment (bbirchnz) [reply src/app/app-winui3/InputSettingsPage.idl:33]: clang-format is picking them up I believe. I'll disable.

File: src/app/app-winui3/InputSettingsPage.xaml.cpp (hunk 1 above)
Comment (bbirchnz) [reply src/app/app-winui3/InputSettingsPage.xaml.cpp:49]: Would this matter if in future I let this be changed in a tab? e.g. having a drawing style/colour/something as a standard part of the header of a TabWithDoodles?

Comment (bbirchnz) [reply src/app/app-winui3/InputSettingsPage.xaml.cpp:49]: If I change bindings in XAML to "OneWay" then changes in the UI don't get applied, and I return to the UI are at their previous value. I must be misunderstanding something here sorry.

File: src/app/app-common/tabs/TabWithDoodles.cpp [hunk 7]
```diff
@@ -97,11 +94,20 @@ void TabWithDoodles::FlushCursorEvents() {
       const bool erasing = event.mButtons & ~1;
 
       const auto scale = mDrawings.at(pageIndex).mScale;
+
+      auto ds = GetKneeboardState()->GetDoodleSettings();
+      auto draw_radius = ds.drawRadius;
```
Comment (fredemmott) [inline src/app/app-common/tabs/TabWithDoodles.cpp:99]: sorry for the nit again: drawRadius, eraseRadius, fixedDraw, fixedErase

File: src/app/app-winui3/InputDeviceUIData.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (fredemmott) [reply src/app/app-winui3/InputDeviceUIData.cpp:65]: `git checkout master InputDeviceUIData.cpp` will get you the previous version, which can then be added to your commit - replacing 'master' with the first commit in your branch that you didn't author

File: src/app/app-winui3/InputSettingsPage.xaml [hunk 8]
```diff
@@ -1,88 +1,54 @@
...
+      <StackPanel Grid.Row="0">
+        <TextBlock Style="{StaticResource TitleTextBlockStyle}" Text="Input Settings" Margin="0,0,0,8" />
+        <TextBlock Text="Pen" Style="{StaticResource SubtitleTextBlockStyle}" Margin="0,0,0,16" />
+        <StackPanel Background="{ThemeResource CardBackgroundFillColorDefaultBrush}" BorderBrush="{ThemeResource CardStrokeColorDefaultBrush}" CornerRadius="4" BorderThickness="1" Padding="8" Margin="0,0,0,8">
+          <Slider Header="Write Radius" Value="{x:Bind WriteRadius, Mode=TwoWay}" TickPlacement="Outside" TickFrequency="10" StepFrequency="1" Maximum="100" Minimum="1.0" />
+          <ToggleSwitch Header="Fixed Write Radius" OnContent="On" OffContent="Off" IsOn="{x:Bind FixedWriteRadius, Mode=TwoWay}" />
+          <Slider Header="Erase Radius" Value="{x:Bind EraseRadius, Mode=TwoWay}" TickPlacement="Outside" TickFrequency="10" StepFrequency="1" Maximum="200" Minimum="1.0" />
+          <ToggleSwitch Header="Fixed Erase Radius" OnContent="On" OffContent="Off" IsOn="{x:Bind FixedEraseRadius, Mode=TwoWay}" />
+        </StackPanel>
+      </StackPanel>
```
[the remark this answers is not in the document]
Comment (fredemmott) [reply src/app/app-winui3/InputSettingsPage.xaml:112]: From discord discussion - can you rebase on https://github.com/fredemmott/OpenKneeboard/commit/a507cecb2b75ec7163bc6cd0311507627fe48f34 ?

File: src/app/app-common/tabs/include/OpenKneeboard/Tab.h [hunk 9]
```diff
@@ -39,9 +39,19 @@ class Tab {
   RenderPage(ID2D1DeviceContext*, uint16_t pageIndex, const D2D1_RECT_F& rect)
     = 0;
 
+  void SetKneeboardState(KneeboardState* kbs_ptr) {
+    pKbs = kbs_ptr;
+  }
+  KneeboardState* GetKneeboardState() {
+    return pKbs;
+  }
```
[the remark this answers is not in the document]
Comment (fredemmott) [reply src/app/app-common/tabs/include/OpenKneeboard/Tab.h:47]: From discord discussion, still todo

File: src/app/app-winui3/InputDeviceUIData.h [hunk 10]
```diff
@@ -51,6 +51,8 @@ struct TabletInputDeviceUIData
   uint8_t Orientation();
   void Orientation(uint8_t value);
 
+
+
```
Comment (fredemmott) [inline src/app/app-winui3/InputDeviceUIData.h:55]: nit: revert these

File: src/app/app-winui3/InputSettingsPage.xaml [hunk 11]
```diff
@@ -71,6 +71,44 @@
...
+        <Slider 
+          Header="Write Radius" 
+          Value="{x:Bind WriteRadius, Mode=TwoWay}" 
+          TickPlacement="Outside" 
+          TickFrequency="10" 
+          StepFrequency="1" 
+          Maximum="100" 
+          Minimum="1.0" />
+        <ToggleSwitch 
+          Header="Fixed Write Radius"
```
Comment (fredemmott) [inline src/app/app-winui3/InputSettingsPage.xaml:94]: Perhaps invert these (just in the UI), and relabel as something like "Pressure-sensitive pen stroke size" + "Pressure-sensitive eraser size" ?

**Review by fredemmott — CHANGES_REQUESTED** [review]
Pretty much there, only requesting changes for the ->SaveSettings() call :)

**Review by fredemmott — APPROVED** [review]

**Pull request merged**